src/coreclr/src/jit/lsra.cpp

```diff
@@ -2994,291 +2989,584 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
+        regMaskTP lowestRegOrderBit = RBM_NONE;
+        for (regMaskTP regOrderCandidates = selector.candidates; regOrderCandidates != RBM_NONE;)
+        {
+            regMaskTP regOrderCandidateBit = genFindLowestBit(regOrderCandidates);
+            regOrderCandidates &= ~regOrderCandidateBit;
+            regNumber regOrderCandidateRegNum = genRegNumFromMask(regOrderCandidateBit);
+            unsigned  thisRegOrder            = getRegisterRecord(regOrderCandidateRegNum)->regOrder;
+            if (thisRegOrder < lowestRegOrder)
+            {
+                lowestRegOrder    = thisRegOrder;
+                lowestRegOrderBit = regOrderCandidateBit;
+            }
         }
-        assignPhysReg(availablePhysRegInterval, currentInterval);
-        foundReg                        = availablePhysRegInterval->regNum;
-        regMaskTP foundRegMask          = genRegMask(foundReg);
-        refPosition->registerAssignment = foundRegMask;
+        assert(lowestRegOrderBit != RBM_NONE);
+        found = selector.applySingleRegSelection(REG_ORDER, lowestRegOrderBit);
     }
 
+    // The set of registers with the lowest spill weight.
+    regMaskTP lowestCostSpillSet = RBM_NONE;
+    // Apply the SPILL_COST heuristic and eliminate regs that can't be spilled.
+    if (!found)
+    {
+        // The spill weight for 'refPosition' (the one we're allocating now).
+        float thisSpillWeight = getWeight(refPosition);
+        // The  spill weight for the best candidate we've found so far.
+        float bestSpillWeight = FloatingPointUtils::infinite_float();
+        // True if we found registers with lower spill weight than this refPosition.
+        bool foundLowerSpillWeight = false;
+
+        for (regMaskTP spillCandidates = selector.candidates; spillCandidates != RBM_NONE;)
+        {
+            regMaskTP spillCandidateBit = genFindLowestBit(spillCandidates);
+            spillCandidates &= ~spillCandidateBit;
+            regNumber  spillCandidateRegNum    = genRegNumFromMask(spillCandidateBit);
+            RegRecord* spillCandidateRegRecord = &physRegs[spillCandidateRegNum];
+
+            // Can and should the interval in this register be spilled for this one,
+            // if we don't find a better alternative?
+            if ((getNextIntervalRefLocation(spillCandidateRegNum ARM_ARG(regType)) == currentLocation) &&
+                !spillCandidateRegRecord->assignedInterval->getNextRefPosition()->RegOptional())
+            {
+                continue;
+            }
+            if (!isSpillCandidate(currentInterval, refPosition, spillCandidateRegRecord))
+            {
+                continue;
+            }
+
+            float currentSpillWeight = spillCost[spillCandidateRegNum];
+#ifdef TARGET_ARM
+            if (currentInterval->registerType == TYP_DOUBLE)
+            {
+                currentSpillWeight = max(currentSpillWeight, spillCost[REG_NEXT(spillCandidateRegNum)]);
```

Comment:
I agree that it would be more consistent, but the fact that the location heuristics are (in some places) using the location of the second register, rather than the min of the two, is, I believe, just an accident.